Sources/TuistAutomation/Device/DeviceController.swift

```diff
@@ -0,0 +1,168 @@
+                    "/usr/bin/xcrun", "devicectl",
+                    "list", "devices",
+                    "--json-output", devicesListOutputPath.pathString,
+                ]
+            )
+            .concatenatedString()
+
+            let deviceList = try JSONDecoder().decode(
+                DeviceList.self,
+                from: try await fileSystem.readFile(at: devicesListOutputPath)
+            )
+
+            return deviceList.result.devices
+                .map(PhysicalDevice.init)
+        }
+    }
+
+    public func installApp(
+        at path: AbsolutePath,
+        device: PhysicalDevice
+    ) async throws {
+        logger.debug("Installing app at \(path) on simulator device with id \(device.id)")
+        do {
+            _ = try await commandRunner.run(
+                arguments: [
+                    "/usr/bin/xcrun", "devicectl",
+                    "device", "install", "app",
+                    "--device", device.id,
+                    path.pathString,
+                ]
+            )
+            .concatenatedString()
+        } catch let error as CommandError {
+            if case let .terminated(_, stderr) = error, stderr.contains("ApplicationVerificationFailed") {
+                throw DeviceControllerError.applicationVerificationFailed
+            } else {
+                throw error
+            }
+        }
+    }
+
+    public func launchApp(
+        bundleId: String,
+        device: PhysicalDevice
+    ) async throws {
+        logger.debug("Launching app with bundle id \(bundleId) on a physical device with id \(device.id)")
+        _ = try await commandRunner.run(
+            arguments: [
+                "/usr/bin/xcrun", "devicectl",
+                "device", "process", "launch",
+                "--device", device.id,
+                bundleId,
+            ]
+        )
+        .concatenatedString()
+    }
+}
+
+private struct DeviceList: Codable {
+    let result: Result
+
+    struct Result: Codable {
+        let devices: [Device]
+
+        struct Device: Codable {
+            let connectionProperties: ConnectionProperties
+            let deviceProperties: DeviceProperties
+            let hardwareProperties: HardwareProperties
+            let identifier: String
+
+            struct ConnectionProperties: Codable {
+                let transportType: String?
+            }
+
+            struct DeviceProperties: Codable {
+                let name: String
+                let osVersionNumber: String
+            }
+
+            struct HardwareProperties: Codable {
+                enum Platform: String, Codable {
+                    case iOS
+                    case tvOS
+                    case watchOS
+                    case visionOS
+                }
+
+                let udid: String
+                let platform: Platform
+            }
+        }
+    }
+}
+
+extension PhysicalDevice {
+    fileprivate init(_ device: DeviceList.Result.Device) {
```

Comment:
Have you considered working directly with `Device` instead of mapping that to `PhysicalDevice`?

Reply:
I'd rather map this to a struct we own instead of taking 1-1 the representation they give us. It's similar to mapping networking models to your own. You have better control over the model – and in the future, it's very likely we'd support Android and we would ideally have one `PhysicalDevice` model for both Android and iOS.